src/tree/updater_quantile_hist.cc

```diff
@@ -574,15 +574,55 @@ void QuantileHistMaker::Builder::InitData(const GHistIndexMatrix& gmat,
       CHECK_EQ(param_.sampling_method, TrainParam::kUniform)
         << "Only uniform sampling is supported, "
         << "gradient-based sampling is only support by GPU Hist.";
-      std::bernoulli_distribution coin_flip(param_.subsample);
       auto& rnd = common::GlobalRandom();
+#if XGBOOST_CUSTOMIZE_GLOBAL_PRNG
+      std::bernoulli_distribution coin_flip(param_.subsample);
       size_t j = 0;
       for (size_t i = 0; i < info.num_row_; ++i) {
         if (gpair[i].GetHess() >= 0.0f && coin_flip(rnd)) {
           p_row_indices[j++] = i;
         }
       }
       row_indices.resize(j);
+#else
+      const size_t nthread = this->nthread_;
+      std::vector<size_t> row_offsets(nthread, 0);
+      /* usage of mt19937_64 give 2x speed up for subsampling */
+      std::vector<std::mt19937> rnds(nthread);
+      /* create engine for each thread */
+      for (std::mt19937& r : rnds) {
+        r = rnd;
+      }
+      const size_t discard_size = info.num_row_ / nthread;
+      #pragma omp parallel num_threads(nthread)
```

Comment:
Overall LGTM, but just to improve - better to move this piece of code to a separate function. Because InitData becomes massive, especially with introducing conditional building.

Maybe it's paranoia, but did you check single-thread perf? Is it the same as before?

Reply:
Thanks. Fixed.
single-thread was checked for santander benchmark (but with 1000 n_iter).
309.4s vs 309.3s
And there should not be any changes for single-thread as seems no additional work was introduced :)